Add Immix GC

wenyuzhao opened the pull request.

wenyuzhao commented on src/memory_manager.rs:
```diff
@@ -185,6 +185,7 @@ pub fn start_worker<VM: VMBinding>(
 /// * `tls`: The thread that wants to enable the collection. This value will be passed back to the VM in
 ///   Collection::spawn_worker_thread() so that the VM knows the context.
 pub fn enable_collection<VM: VMBinding>(mmtk: &'static MMTK<VM>, tls: VMThread) {
+    mmtk.plan.pre_worker_spawn(mmtk);
```
gc_init happens before enable_collection. The number of worker threads in the scheduler is not set up in gc_init. So moving defrag init code to gc_init won't work.

qinsoon replied:
My opnion:
1. You don't actually need to get the worker number from the scheduler. You can get it from `Options.threads`. Each plan has a reference to `Options`.
2. I don't mind adding a new hook if it is necessary. E.g. we do want to do something exactly before spawning workers but after gc_init. However, for this case (`initialize_defrag()`), I think it is okay to do it in gc_init.

wenyuzhao replied:
This hook is removed

qinsoon reviewed (commented):
These are some high-level reviews before I would go for details:
* Generate work packets: it seems you try generate work packets based on the number of worker threads, however I don't think this is desired. Generally we want to generate a fixed number of packets, each of which contains a considerable amount of work, and we let workers to compete and finish them. You do not need to try to balance the number of packets and try get one packet for each thread, it's the scheduler's job. With this removed, I don't think you need a reference from the immix policy to the scheduler any more. You can take a look at how Marksweep with Kunal's changes generates sweep tasks for the chunks. 
* Mark bit: you need to use the mark bit from ObjectModel. It could be in header or side.
* Chunk map: it is basically side metadata (1 bit per chunk). Is there any reason you are not using our side metadata implementation for it? It is nice to see you implemented those iterators for chunks. Iteartion in our side metadata is missing though.
* Unsafe code: this PR includes some unnecessary unsafe code, e.g. the pattern `UnsafeCell<CommonSpace>` which has been removed from other spaces. Please check each case. If you really need unsafe, please add a comment to justify the use of unsafe is appropriate and correct.
* Address calculation: for most calculation based on addresses, we have implemented those ops for `Address`. No need to turn an address to usize for calculation.

wks commented:
I think the plan does NOT need to know the number of workers to initialise. i.e. the `pre_worker_spawn` is redundant, and we should use a more conventional message queue mechanism.

This function in Immix initialises the histogram with the number of workers, and the reason is related to multi-threaded GC.  Immix calculates the histogram at the end of each GC, and use it to estimate the defragment threshold during the next GC.  Because sweeping is multi-threaded, each GC worker constructs its own histogram (https://github.com/mmtk/mmtk-core/pull/382/files#diff-68ba28c9c7607f62ad43d87ef1ad6cccccf9b990c90c10b9a41e8bf53ec7344cR135) during sweeping, and the next GC aggregates all the histograms (https://github.com/mmtk/mmtk-core/pull/382/files#diff-49961236d9b4ffb9b6ca6f3fc8185727c117524f83ddc8689ab095227685147aR169) and work out the threshold.  The two-dimensional Vec `spill_mark_histograms` here serves as a **communication channel between GC workers and the next GC**.

I consider it a map-reduce-style design pattern. All we need is a place to aggregate the results of GC works.  That does not necessarily need to be a fixed-length array.  Any container should work.  Consider the following pseudo-code:

```rust
global:
  type Histogram = Vec<usize>;
  let mark_histograms: Mutex<Vec<Histogram>> = Mutex::new(vec![]);

worker:
  let histogram: Histogram = work_out_the_histogram();
  {
    let lock_guard = mark_histograms.lock().unwrap();
    lock_guard.push(histogram);
  }

establish_defrag_spill_threshold:
  let histograms: Vec<Histogram> = {
    let lock_guard = mark_histograms.lock().unwrap();
    lock_guard.drain().collect::<_>()
  };
  let threshold = work_out_the_threshold(&histograms);
```

If the global `mark_histograms` is a dynamically-sized vector, it does not need to know the number of workers.  Of course this may introduce some synchronisation overhead, but I don't think it is significant.  Moreover, it makes the system more flexible, and we can even increase or decrease the number of workers at runtime on demand.

wenyuzhao commented:
@Yi for your high-level suggestions,

I've fixed mark bit, chunk map and address calculation part. I've also addressed and removed unsafe code as much as possible. The remaining unsafe code are mostly be commented with safety explanations.

For work-packets part, as @wks mentioned, the histogram calculation is indeed a map-reduce pattern. So I used Kunshan's solution and completely removed the `pre_worker_spawn` hook (the hook is only used to construct the histograms, so it is not needed any more).

I've removed the chunk sweeping work balancing code and created a sweeping  work packet for each chunk. But I still need to hold a reference to scheduler because the sweeping work can only be created after defrag.prepare() is done in ImmixSpace::prepare(). So I cannot create these work packets in Immix plan `schedule_collection` function. BTW, keeping a cyclic weak reference is not that bad... its not an anti-rust pattern.

qinsoon commented:
> BTW, keeping a cyclic weak reference is not that bad... its not an anti-rust pattern.

@wenyuzhao For your case, it is not cyclic reference, right? The immix policy has a reference `Scheduler`, but the `Scheduler` does not have references for the policy. As for cyclic reference, it is generally considered an anti-pattern for programming (not only for Rust) [1, 2].
[1] https://en.wikipedia.org/wiki/Circular_dependency
[2] https://softwareengineering.stackexchange.com/questions/11856/whats-wrong-with-circular-references

qinsoon commented on src/util/alloc/immix_allocator.rs:
```diff
@@ -0,0 +1,233 @@
+    /// approximation to bytes allocated
+    line_use_count: usize,
+    /// Line mark table
+    mark_table: Address,
```
This is unused.

wenyuzhao replied:
Fixed. As well as other unused fields.

wks reviewed (commented):
Here are some comments about

-   `spill_avail_histograms`
-   potential overflow related to the sizes of blocks and lines.

wenyuzhao commented on src/plan/immix/gc_work.rs:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    #[inline(always)]
+    fn process_edge(&mut self, slot: Address) {
```
Yes it's duplicate code. Removed.

wenyuzhao commented on src/plan/immix/global.rs:
```diff
@@ -0,0 +1,158 @@
+    pub common: CommonPlan<VM>,
+}
+
+unsafe impl<VM: VMBinding> Sync for Immix<VM> {}
```
Removed.

wenyuzhao commented on src/plan/immix/gc_work.rs:
```diff
@@ -0,0 +1,198 @@
+    }
+}
+
+pub struct ImmixProcessEdges<VM: VMBinding> {
```
Done. I did not separate into two work packets. I followed Kunshan's idea instead to add a constant generic parameter to the type to remove dynamic branching.

wenyuzhao commented on src/plan/immix/gc_work.rs:
```diff
@@ -0,0 +1,198 @@
+
+    /// Trace objects without evacuation.
+    #[inline(always)]
+    fn fast_process_edge(&mut self, slot: Address) {
```
No. This calls `fast_trace_object` not `trace_object`. Otherwise it's the same as `process_edge`.

qinsoon replied:
I meant if you separate the two types of packets, then this process edge is the same as the default process_edge. But it doesn't matter if you are not using two different types for those packets.

wenyuzhao commented on src/policy/immix/block.rs:
```diff
@@ -0,0 +1,299 @@
+    }
+    /// result = block_address + count * block_size
+    #[inline(always)]
+    fn forward_checked(start: Self, count: usize) -> Option<Self> {
```
Fixed. I added overflow/underflow checks and return None if this happens.

wenyuzhao commented on src/policy/immix/chunk.rs:
```diff
@@ -0,0 +1,259 @@
+                .blocks()
+                .filter(|block| block.get_state() != BlockState::Unallocated)
+            {
+                // Calculate number of marked lines and holes.
```
Done.

wenyuzhao commented on src/policy/immix/block.rs:
```diff
@@ -0,0 +1,299 @@
+        unsafe { &*start.to_ptr() }
+    }
+
+    const MARK_UNALLOCATED: u8 = 0;
```
Done

wenyuzhao commented on src/policy/immix/defrag.rs:
```diff
@@ -0,0 +1,194 @@
+
+    /// Get the number of defrag headroom pages.
+    pub fn defrag_headroom_pages<VM: VMBinding>(&self, space: &ImmixSpace<VM>) -> usize {
+        space.get_page_resource().reserved_pages() * 2 / 100
```
Sure. I've replaced `2` by a percentage constant but still leaves 100 here. Moving `2/100` to a constant will lead to floating point multiplication which may hit the performance.

wenyuzhao commented on src/policy/immix/immixspace.rs:
```diff
@@ -0,0 +1,629 @@
+        self.get_name()
+    }
+    fn is_live(&self, object: ObjectReference) -> bool {
+        ForwardingWord::is_forwarded::<VM>(object)
```
Fixed. Now checks both `is_marked` and `is_forwarded`

wenyuzhao commented on src/policy/immix/immixspace.rs:
```diff
@@ -0,0 +1,629 @@
+        ForwardingWord::is_forwarded::<VM>(object)
+    }
+    fn is_movable(&self) -> bool {
+        true
```
Fixed

wenyuzhao commented on src/policy/immix/immixspace.rs:
```diff
@@ -0,0 +1,629 @@
+    /// Object mark state
+    mark_state: usize,
+    /// Work packet scheduler
+    scheduler: Weak<MMTkScheduler<VM>>,
```
Fixed. I thought this is a cyclic reference but it's actually not. I've changed it back to `Arc`

wenyuzhao commented on src/policy/immix/mod.rs:
```diff
@@ -0,0 +1,27 @@
+pub use immixspace::*;
+
+/// Mark/sweep memory for block-level only
+pub const BLOCK_ONLY: bool = false;
```
This feature turns immix into a simple mark-region collector with one-level region hierarchy. This is useful for doing experiments on immix. e.g. I'm using it to experiment with concurrent marking -- enabling this feature gives me a super simple mark-region + non-moving GC and is much easier for debugging.

wenyuzhao commented on src/policy/immix/block.rs:
```diff
@@ -0,0 +1,299 @@
+    const MARK_MARKED: u8 = u8::MAX - 1;
+
+    #[inline(always)]
+    fn mark_byte(&self) -> &AtomicU8 {
```
Fixed

wenyuzhao commented on src/policy/immix/block.rs:
```diff
@@ -0,0 +1,299 @@
+    const DEFRAG_SOURCE_STATE: u8 = u8::MAX;
+
+    #[inline(always)]
+    fn defrag_byte(&self) -> &AtomicU8 {
```
Fixed

wenyuzhao commented on src/policy/immix/defrag.rs:
```diff
@@ -0,0 +1,194 @@
+            && (emergency_collection
+                || (collection_attempts > 1)
+                || !exhausted_reusable_space
+                || Self::DEFRAG_STRESS);
```
Fixed

wenyuzhao commented on src/policy/immix/defrag.rs:
```diff
@@ -0,0 +1,194 @@
+        }
+    }
+
+    /// Release work. Should be called in ImmixSpace::prepare.
```
Fixed

wenyuzhao commented on src/policy/immix/defrag.rs:
```diff
@@ -0,0 +1,194 @@
+    /// Update available_clean_pages_for_defrag counter when a clean block is allocated.
+    pub fn notify_new_clean_block(&self, copy: bool) {
+        if copy {
+            let available_clean_pages_for_defrag = self
```
Fixed

qinsoon replied:
The update isn't correct. https://github.com/mmtk/mmtk-core/commit/46dc113a03bb60b8b44dcdf5490b4087647be0c2#r53964467

wks commented on src/policy/immix/block.rs:
```diff
@@ -0,0 +1,299 @@
+    #[inline(always)]
+    pub fn set_holes(&self, holes: usize) {
+        self.defrag_byte().store(holes as _, Ordering::Release);
+    }
```
Note that the problem is not solved after switching to `side_metadata::store_atomic`. Now `DEFRAG_STATE_TABLE.log_num_of_bits` shall be sufficient for the block/line size configuration.

wenyuzhao replied:
Done. I've put assertions in `ImmixSpace::init` to check the potential overflow. If this actually happens due to the change of block/line sizes, then the whole metadata layout should be updated as well.

wenyuzhao commented on src/policy/immix/block.rs:
```diff
@@ -0,0 +1,299 @@
+/// A non-block single-linked list to store blocks.
+#[derive(Default)]
+pub struct BlockList {
+    queue: RwLock<SegQueue<Block>>,
```
Fixed. Using `Mutex<Vec<Block>>` now.

wenyuzhao commented on src/policy/immix/chunk.rs:
```diff
@@ -0,0 +1,259 @@
+            return;
+        }
+        // Update alloc byte
+        unsafe { side_metadata::store(&Self::ALLOC_TABLE, chunk.start(), state as u8 as _) };
```
Yes. Chunk mark byte updating only happens at two places: 1. Set a chunk to live when allocating a new block. 2. Set a chunk to dead when sweeping blocks. These two operations cannot happen at the same time. So they can always use unconditional non-atomic stores.

wenyuzhao commented on src/policy/immix/chunk.rs:
```diff
@@ -0,0 +1,259 @@
+
+    /// Get chunk state
+    pub fn get(&self, chunk: Chunk) -> ChunkState {
+        let byte = unsafe { side_metadata::load(&Self::ALLOC_TABLE, chunk.start()) as u8 };
```
Same as above.

wenyuzhao commented on src/policy/immix/immixspace.rs:
```diff
@@ -0,0 +1,629 @@
+    /// Line mark state in previous GC
+    line_unavail_state: AtomicU8,
+    /// Is in a GC?
+    in_collection: AtomicBool,
```
Fixed

wenyuzhao commented on src/policy/immix/defrag.rs:
```diff
@@ -0,0 +1,194 @@
+    /// A list of completed mark histograms reported by workers
+    pub mark_histograms: Mutex<Vec<MarkHistogram>>,
+    /// Summarised histograms
+    spill_avail_histograms: Vec<AtomicUsize>,
```
Done. I renamed `MarkHistogram` to `Histogram` and turn `spill_avail_histograms` into a local variable, with type `Histogram`.

wenyuzhao commented on src/policy/immix/chunk.rs:
```diff
@@ -0,0 +1,259 @@
+                    if marked_lines != Block::LINES {
+                        // There are holes. Mark the block as reusable.
+                        block.set_state(BlockState::Reusable {
+                            unavailable_lines: marked_lines as _,
```
Assertions in `ImmixSpace::init` covers this as well.

wenyuzhao commented on src/policy/immix/defrag.rs:
```diff
@@ -0,0 +1,194 @@
+                .load(Ordering::Acquire)
+                << (LOG_BYTES_IN_PAGE as usize - Line::LOG_BYTES));
+
+        let mut required_lines = 0isize;
```
Done.

wenyuzhao commented on src/policy/immix/immixspace.rs:
```diff
@@ -0,0 +1,629 @@
+            // Object mark table will be cleared by `PrepareBlockState` before each GC.
+            //
+            // Note: It is incorrect to flip matk bit between 0 and 1 and remove
+            // the mark-table zeroing step. Because openjdk does not call post_alloc to set up
```
Actually this comment is not correct... we always need to clear the mark table for each GC, if they are on the side. I've removed this comment.

wenyuzhao replied:
This does not related to whether `post_alloc` is implemented or not.

wenyuzhao commented on src/policy/immix/immixspace.rs:
```diff
@@ -0,0 +1,629 @@
+    fn delta_mark_state(state: u8) -> u8 {
+        debug_assert!(!VM::VMObjectModel::LOCAL_MARK_BIT_SPEC.is_on_side());
+        let mut rtn = state;
+        loop {
```
The `VMLocalMarkBitSpec` only reserve a bit for each object for marking. So it cannot be used to implement cyclic mark state anymore. I've removed this code and put "unimplemented" in `ImmixSpace::prepare` if `LOCAL_MARK_BIT_SPEC` is in header.

I'll create an issue about this.

wenyuzhao commented on src/policy/immix/immixspace.rs:
```diff
@@ -0,0 +1,629 @@
+    /// Defrag utilities
+    pub(super) defrag: Defrag,
+    /// Object mark state
+    mark_state: usize,
```
Fixed. use u8 now.

wenyuzhao commented on src/policy/immix/immixspace.rs:
```diff
@@ -0,0 +1,629 @@
+}
+
+/// A transitive closure visitor to collect all the edges of an object.
+pub struct ObjectsClosure<'a, E: ProcessEdgesWork>(
```
Should I move this to `crate::util` so bindings can use this?

qinsoon replied:
If it is not specific to immix, you can move it to somewhere else. I am not concerning where you move it to. You can put it to `plan` - we have transitive closure there. You can put it to `util` or `scheduler::gc_work`.

wenyuzhao replied:
Done. I moved it to plan::transitive_closure.

qinsoon commented on src/policy/immix/block.rs:
```diff
@@ -0,0 +1,299 @@
+    pub const DEFRAG_STATE_TABLE: SideMetadataSpec = SideMetadataSpec {
+        is_global: false,
+        offset: if super::BLOCK_ONLY {
+            LOCAL_SIDE_METADATA_BASE_OFFSET
```
I think you need to put the immix specs after the current local specs. See here:
https://github.com/mmtk/mmtk-core/blob/17b2537e9a3bcc8ead34c4288d5b21efdca80404/src/util/metadata/side_metadata/global.rs#L234-L242

wenyuzhao replied:
I think I only need to pass the used local side metadata to CommonSpace::side_metadata_specs. This is what other spaces do. The offset for each spec is calculated using `SideMetadataOffset::layout_after`on definition.

wenyuzhao commented on src/policy/immix/block.rs:
```diff
@@ -0,0 +1,299 @@
+    pub const LINES: usize = 1 << Self::LOG_LINES;
+
+    /// Block defrag state table (side)
+    pub const DEFRAG_STATE_TABLE: SideMetadataSpec = SideMetadataSpec {
```
Fixed. But I'm having a looooong side spec chain now, constructed and connect using `SideMetadataOffset::layout_after`:

`LOCAL_SIDE_METADATA_BASE_OFFSET`
-> `malloc ALLOC_SIDE_METADATA_SPEC`-> `malloc ACTIVE_PAGE_METADATA_SPEC`
-> `immix Line::MARK_TABLE` -> `immix Block::DEFRAG_STATE_TABLE` -> `immix Block::MARK_TABLE` -> `immix ChunkMap::ALLOC_TABLE`
-> `LOCAL_SIDE_METADATA_VM_BASE_OFFSET`

This looks terrible. Immix space local metadata should not depend on Malloc space specs....

wenyuzhao replied:
I had a chat with Steve. We should fix this in the future (not in this PR). Local side metadata spec in different spaces should not depend on each other.

qinsoon replied:
I have created an issue for this. https://github.com/mmtk/mmtk-core/issues/398

qinsoon commented on src/util/metadata/side_metadata/constants.rs:
```diff
@@ -103,6 +103,5 @@ pub const GLOBAL_SIDE_METADATA_VM_BASE_OFFSET: SideMetadataOffset =
-pub const LOCAL_SIDE_METADATA_VM_BASE_OFFSET: SideMetadataOffset = SideMetadataOffset::layout_after(
-    &crate::policy::mallocspace::metadata::ACTIVE_PAGE_METADATA_SPEC,
-);
+pub const LOCAL_SIDE_METADATA_VM_BASE_OFFSET: SideMetadataOffset =
```
Please update the comments above this constant, add the policy specific specs you added in the PR.

wenyuzhao replied:
Done.

wenyuzhao commented on src/policy/immix/block.rs:
```diff
@@ -0,0 +1,299 @@
+    /// Get the address range of the block's line mark table.
+    #[allow(clippy::assertions_on_constants)]
+    #[inline(always)]
+    pub fn line_mark_table(&self) -> &[AtomicU8; Block::LINES] {
```
Fixed. I added a byte array api to side metadata interface and added the extreme_assertions checks.

qinsoon commented:
I think this PR is ready. I will run binding tests just to make sure it does not break other bindings. If everything is good, we can merge it.

qinsoon commented:
openjdk-binding-test
OPENJDK_BINDING_REF=immix

qinsoon commented:
The JikesRVM binding failed for NoGC. The OpenJDK binding failed at a known issue. 
Can you check the JikesRVM fail? @wenyuzhao

wenyuzhao commented:
jikesrvm-binding-test
JIKESRVM_BINDING_REF=add-immix-allocator

qinsoon commented:
The tests for JikesRVM timed out for 90mins. That seems strange. The tests should take ~50mins.
